// react-charts/lib/utils/HyperResponsive.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

exports.default = HyperResponsive;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

require('javascript-detect-element-resize');

var _Utils = require('../utils/Utils');

var _Utils2 = _interopRequireDefault(_Utils);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }
//


function HyperResponsive(WrappedComponent) {
  return function (_PureComponent) {
    _inherits(HyperResponsive, _PureComponent);

    function HyperResponsive() {
      _classCallCheck(this, HyperResponsive);

      var _this = _possibleConstructorReturn(this, (HyperResponsive.__proto__ || Object.getPrototypeOf(HyperResponsive)).call(this));

      _this.state = {
        ready: false,
        width: 0,
        height: 0
      };
      _this.resize = _this.resize.bind(_this);
      return _this;
    }

    _createClass(HyperResponsive, [{
      key: 'componentWillMount',
      value: function componentWillMount() {
        this.resize = _Utils2.default.throttle(this.resize, 16);
      }
    }, {
      key: 'componentDidMount',
      value: function componentDidMount() {
        if (!this.resizeListener) {
          this.resizeListener = window.addResizeListener(this.el, this.resize);
        }
        this.resize();
      }
    }, {
      key: 'componentWillUnmount',
      value: function componentWillUnmount() {
        this.resizeListener && window.removeResizeListener(this.el, this.resize);
      }
    }, {
      key: 'resize',
      value: function resize(e) {
        this.setState({
          ready: true,
          width: parseInt(window.getComputedStyle(this.el).width),
          height: parseInt(window.getComputedStyle(this.el).height)
        });
      }
    }, {
      key: 'render',
      value: function render() {
        var _this2 = this;

        var _props = this.props,
            style = _props.style,
            rest = _objectWithoutProperties(_props, ['style']);

        var _state = this.state,
            ready = _state.ready,
            width = _state.width,
            height = _state.height;

        return _react2.default.createElement(
          'div',
          {
            className: 'ResponsiveWrapper',
            ref: function ref(el) {
              _this2.el = el;
            },
            style: _extends({
              width: '100%',
              height: '100%'
            }, style)
          },
          ready && _react2.default.createElement(WrappedComponent, _extends({ width: width, height: height }, rest))
        );
      }
    }]);

    return HyperResponsive;
  }(_react.PureComponent);
}
