// react-state/lib/Provider.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

exports.default = function (ComponentToWrap) {
  var _class, _temp;

  var config = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : defaultConfig;

  return _temp = _class = function (_PureComponent) {
    _inherits(Provider, _PureComponent);

    function Provider(props) {
      _classCallCheck(this, Provider);

      var _this = _possibleConstructorReturn(this, (Provider.__proto__ || Object.getPrototypeOf(Provider)).call(this));

      var children = props.children,
          rest = _objectWithoutProperties(props, ['children']);
      // Initialize the store with initial state and props


      _this.store = _extends({}, config.initial, rest);
      _this.subscriptions = [];
      _this.subscribe = _this.subscribe.bind(_this);
      _this.dispatch = _this.dispatch.bind(_this);
      return _this;
    }
    // Define our context key


    _createClass(Provider, [{
      key: 'componentWillReceiveProps',
      value: function componentWillReceiveProps(newProps) {
        // If the component receives new props, merge them into
        // the store and notify subscribers
        this.dispatch(function (state) {
          return _extends({}, state, newProps);
        });
        this.forceUpdate();
      }
    }, {
      key: 'subscribe',
      value: function subscribe(connect) {
        var _this2 = this;

        var meta = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

        var subscription = {
          connect: connect,
          meta: meta
        };
        // Add the subscription
        this.subscriptions.push(subscription);
        // return an unsubscribe function
        return function () {
          _this2.subscriptions = _this2.subscriptions.filter(function (d) {
            return d !== subscription;
          });
        };
      }
    }, {
      key: 'dispatch',
      value: function dispatch(fn) {
        var meta = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

        // When we recieve a dispatch command, build a new version
        // of the store by calling the dispatch function

        // TODO: beforeDispatch
        var oldStore = this.store;
        var newStore = fn(oldStore);
        // TODO: middleware
        this.store = newStore;
        this.subscriptions.forEach(function (subscription) {
          var shouldNotify = true;
          if (subscription.meta.filter) {
            shouldNotify = subscription.meta.filter(oldStore, newStore, meta);
          }
          shouldNotify && subscription.connect();
        });
        this.forceUpdate();
      }
    }, {
      key: 'getChildContext',
      value: function getChildContext() {
        var _this3 = this;

        return {
          reactState: {
            getStore: function getStore() {
              return _this3.store;
            },
            subscribe: this.subscribe,
            dispatch: this.dispatch
          }
        };
      }
    }, {
      key: 'render',
      value: function render() {
        return _react2.default.createElement(
          ComponentToWrap,
          this.store,
          this.props.children
        );
      }
    }]);

    return Provider;
  }(_react.PureComponent), _class.childContextTypes = {
    reactState: _propTypes2.default.object.isRequired
  }, _temp;
};

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var defaultConfig = {
  initial: {}
};
