// react-charts/lib/components/Brush.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactState = require('react-state');

var _Selectors = require('../utils/Selectors');

var _Selectors2 = _interopRequireDefault(_Selectors);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }
//


// import Rectangle from '../primitives/Rectangle'

var Brush = function (_PureComponent) {
  _inherits(Brush, _PureComponent);

  function Brush() {
    _classCallCheck(this, Brush);

    return _possibleConstructorReturn(this, (Brush.__proto__ || Object.getPrototypeOf(Brush)).apply(this, arguments));
  }

  _createClass(Brush, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      var _props = this.props,
          onSelect = _props.onSelect,
          cursor = _props.cursor,
          primaryAxis = _props.primaryAxis;

      if (this.props.cursor && nextProps.cursor.released !== this.props.cursor.released) {
        onSelect({
          cursor: nextProps.cursor.released,
          start: primaryAxis.scale.invert(cursor.sourceX),
          end: primaryAxis.scale.invert(cursor.x)
        });
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          _props2$cursor = _props2.cursor,
          cursor = _props2$cursor === undefined ? {} : _props2$cursor,
          offset = _props2.offset,
          gridX = _props2.gridX,
          gridY = _props2.gridY,
          gridHeight = _props2.gridHeight;


      return _react2.default.createElement(
        'div',
        {
          className: 'Brush',
          style: {
            pointerEvents: 'none',
            position: 'absolute',
            left: offset.left + gridX + 'px',
            top: offset.top + gridY + 'px',
            opacity: !cursor.dragging ? 0 : 1
          }
        },
        _react2.default.createElement('div', {
          style: {
            position: 'absolute',
            left: Math.min(cursor.x, cursor.sourceX) + 'px',
            width: Math.abs(cursor.x - cursor.sourceX) + 'px',
            height: gridHeight + 'px',
            background: 'rgba(0,0,0,.2)'
          }
        })
      );
    }
  }]);

  return Brush;
}(_react.PureComponent);

Brush.defaultProps = {
  onSelect: function onSelect() {}
};
exports.default = (0, _reactState.Connect)(function (state) {
  var selectors = {
    primaryAxis: _Selectors2.default.primaryAxis(),
    secondaryAxis: _Selectors2.default.secondaryAxis(),
    offset: _Selectors2.default.offset(),
    gridHeight: _Selectors2.default.gridHeight(),
    gridX: _Selectors2.default.gridX(),
    gridY: _Selectors2.default.gridY()
  };
  return function (state) {
    return {
      primaryAxis: selectors.primaryAxis(state),
      secondaryAxis: selectors.secondaryAxis(state),
      cursor: state.cursor,
      offset: selectors.offset(state),
      gridHeight: selectors.gridHeight(state),
      gridX: selectors.gridX(state),
      gridY: selectors.gridY(state)
    };
  };
}, {
  statics: {
    isHTML: true
  }
})(Brush);
